### ISSUE TITLE
Xwayland is delaying commits for vsynced clients

### ISSUE THREAD

**Issue by Plagman** [Plagman/gamescope#25]
Looking at gpuvis captures for vsynced games running is xwayland, unwanted things are happening. Xwayland is decoupled from the thing actually handling the output, so I'm not sure it ought having a say. Ideally Xwayland would just forward whether the client wants vsync to the compositor, and the compositor could pass back enough information to have the right behavior instead.

**Comment by subdiff** [comment]
Currently XWayland works like this for vsynced games:
1.  Flips get marked with [sync_flip](https://gitlab.freedesktop.org/xorg/xserver/blob/435d41d5ffe5467f43225b59411ca9f9a05c10a6/hw/xwayland/xwayland-present.c#L447).
1. On frame callback the MSC is [bumped](https://gitlab.freedesktop.org/xorg/xserver/blob/435d41d5ffe5467f43225b59411ca9f9a05c10a6/hw/xwayland/xwayland-present.c#L264).
1. The flip gets [marked](https://gitlab.freedesktop.org/xorg/xserver/blob/435d41d5ffe5467f43225b59411ca9f9a05c10a6/hw/xwayland/xwayland-present.c#L215) as done.

This has the following issue: The frame callback is not received by XWayland when the actual frame was presented but potentially later on when the compositor informs clients when there is ["a good time to start drawing the next one"](https://wayland.freedesktop.org/docs/html/apa.html#protocol-spec-wl_surface-request-frame).

I'm currently looking into replacing this with presentation-time protocol: [Task](https://gitlab.freedesktop.org/xorg/xserver/issues/971), [WIP-branch](https://gitlab.freedesktop.org/romangg/xserver/commits/presentation-time)

My current approach is in a first step to decouple XWayland frame stepping from the frame-callback completely and instead directly use the timing values received through presentation-time to drive with an internal timer the presentation (and realign the timer later on).

Second step would then be to maybe reintroduce the frame callback to throttle presenting when the surface is not visible (battery saving). But that depends on how much a difference this actually would make.

Do you have specific games where you noticed this delay?

**Comment by emersion** [comment]
> The frame callback is not received by XWayland when the actual frame was presented but potentially later on when the compositor informs clients when there is "a good time to start drawing the next one".

We send a frame callback as soon as we've received a commit, so this shouldn't be an issue.

Would still be nice to have presentation-time support in Xwayland ofc :)

**Comment by Plagman** [comment]
It's not quite as soon as, so it would still incur a delay for us. I think in general, what I stated back in January is still reasonable: a vsynced commit should be sent ASAP to the compositor without any more delay than a non-vsynced one, and knowledge that this commit/client is currently vsynced should come through another channel.

Has the current situation changed in any way? As a short-term workaround, can we force Xwayland to always act as if vsync is off? gamescope always acts as if vsync is on, so it would be the most efficient path.

**Comment by emersion** [comment]
Adding more trace points, it seems like Xwayland is committing a buffer right after we request it:

- t=0ms: gamescope calls `paint_all`
- t=0.12ms: gamescope sends a `wl_surface.frame` event to Xwayland, requesting a new commit
- t=0.20ms: the vulkan WSI in the sekiro process sends a `Present` request to display a new frame
- t=0.55ms: gamescope receives the `wl_surface.commit` event with the new frame

**Comment by emersion** [comment]
OK, it turns out there is some delay, because we actually don't send frame callbacks when the client commits, we send frame callbacks after painting.

Opened an upstream issue: https://gitlab.freedesktop.org/xorg/xserver/-/issues/1150

**Comment by emersion** [comment]
Ref https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/11290

**Comment by Plagman** [comment]
Sounds like discussion is mostly unblocked on that MR and we could proceed forward with disabling the mesa-side waiting based on a driconf setting.

**Comment by emersion** [comment]
The upstream MR has been merged.
